pkg/kubectl/cmd/completion.go

```diff
@@ -308,3 +314,1203 @@ _complete kubectl 2>/dev/null
 	out.Write([]byte(zsh_tail))
 	return nil
 }
+
+func runCompletionFish(out io.Writer, boilerPlate string, kubectl *cobra.Command) error {
+	if len(boilerPlate) == 0 {
+		boilerPlate = defaultBoilerPlate
+	}
+	if _, err := out.Write([]byte(boilerPlate)); err != nil {
+		return err
+	}
+
+	fish_body := `
+function __kubectl_no_command
+    set -l cmd (commandline -poc)
+    if not set -q cmd[2]
+        return 0
+    end
+    return 1
+end
+
+function __kubectl_using_command
+    set cmd (commandline -poc)
+
+    if test (count $cmd) -gt (count $argv)
+        set -e cmd[1]
+        string match -q -- "$argv*" "$cmd"
+        return $status
+    end
+
+    return 1
+end
+
+function __kubectl_using_option
+    set cmd (commandline -poc)
+    set query "("(string join -- "|" $argv)")"
+
+    if test (count $cmd) -gt 1
+        if string match -qr -- $query $cmd[-1]
+            return 0
+        end
+    end
+    return 1
+end
+
+function __kubectl_using_option_value -a option -a value
+    set cmd (commandline -poc)
+
+    if test (count $cmd) -gt 1
+        string match -qr -- $option"[= ]"$value "$cmd"
+        return $status
+    end
+
+    return 1
+end
+
+function __kubectl_using_go_template_format
+    __kubectl_using_option_value output go-template
+    or __kubectl_using_option_value output go-template-file
+    or __kubectl_using_option_value o go-template
+    or __kubectl_using_option_value o go-template-file
+end
+
+function __kubectl_using_resource_type
+    for rt in (__kubectl_resource_types)
+        if __fish_seen_subcommand_from $rt
+            return 0
+        end
+    end
+
+    return 1
+end
+
+function __kubectl_no_pod
+    set cmd (commandline -pc)
+    for i in (__kubectl_pods --no-prefix pods)
+        if string match -q "*$i*" -- $cmd
+            return 1
+        end
+    end
+    return 0
+end
+
+function __kubectl_resource_types
+    echo clusters
```

Comment:
can't these be generated instead of a hard-coded list?

Reply:
yes, with this code:
```
kubectl get 2>&1 | egrep -o '\* [^ ]+' | cut -d' '  -f2
```